File: 2583. Kth Largest Sum in a Binary Tree/Solution.py
from collections import deque
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def kthLargestLevelSum(self, root: Optional[TreeNode], k: int) -> int:
        cola = deque([root])
        levelsSum = []
        while cola:
            size = len(cola)
            sum = 0
            for i in range(size):
                node = cola.popleft()
                sum += node.val
                if node.left:
                    cola.append(node.left)
                if node.right:
                    cola.append(node.right)
            levelsSum.append(sum)
        levelsSum.sort(reverse=True)
        if k <= len(levelsSum):
            return levelsSum[k - 1]

        return -1

File: 2583. Kth Largest Sum in a Binary Tree/Solution.java
//  https://leetcode.com/problems/kth-largest-sum-in-a-binary-tree/?envType=daily-question&envId=2024-10-22
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
public class Solution {
    public long kthLargestLevelSum(TreeNode root, int k) {
        Queue<TreeNode> cola = new LinkedList<>();
        var levelSum = new ArrayList<Long>();
        cola.add(root);
        while (!cola.isEmpty()) {
            int size = cola.size();
            long sum = 0;
            for (int i = 0; i < size; i++) {
                var node = cola.poll();
                sum += node.val;
                if (node.left != null)
                    cola.add(node.left);
                if (node.right != null)
                    cola.add(node.right);
            }
            levelSum.add(sum);
        }
        levelSum.sort(Collections.reverseOrder());
        if (k <= levelSum.size())
            return levelSum.get(k - 1);
        return -1;
    }
}
